### DISCUSSION TITLE
What's the plan for the org.graalvm.sdk:graal-sdk maven artifact?

### DISCUSSION THREAD

**Discussion by zakkak** [oracle/graal#7224] [General] [answered]
https://github.com/oracle/graal/pull/7171 moved package `com.oracle.svm.core.annotate` (which AFAIK is public API) from `graal-sdk.jar` to `nativeimage.jar`.

What are the long term plans for providing access to the said package? Is `org.graalvm.sdk:graal-sdk` going to be maintained or is it getting replaces by some other artifact?

cc @chumer  @fniephaus @jerboaa

**Comment by fniephaus** [comment]
From what I've heard from the team is that we plan to continue ship `org.graalvm.sdk:graal-sdk`, at least for a while, with a dependency on all new artifacts, so depending on the sdk jar will still give you what you're getting today. Maybe @chumer can confirm?

**Reply by jerboaa** [reply]
`graal-sdk.jar` from latest master looks like this:

```
$ jar -tf lib/jvmci/graal-sdk.jar
org/graalvm/sdk/Placeholder.class
module-info.class
```

Clearly, that's not an API jar. Which yields the following question: What's the API (in a maven sense) going forward for GraalVM native image? That API should avoid some bridge artefact, which pulls in other things that provides the real API. Historically, I thought it was `graal-sdk.jar` and that's the reason why we've moved Quarkus to it, but apparently it's no longer the case and we need to move again. Instead of sticking to an old placeholder artefact, shouldn't projects depend on the actual thing that's considered API? It looks like this is going to be `org.graalvm.sdk:nativeimage` going forward (23.1 and better). Is that the case? If not, which is it? In an ideal world that API jar should stay mostly stable, not change frequently if possible.

In a nutshell, we'd like to have Maven coordinates that external projects can depend on and (which remains stable). The packages/classes within it should also be staying mostly stable. Note that those Maven coordinate and API changes tend to have a ripple effect on external projects. Thanks for your help!

**Comment by jerboaa** [comment]
@zakkak Note that according to https://github.com/oracle/graal/pull/4783 and friends, `org.graalvm.nativeimage` is also a package with public API, right?

**Reply by zakkak** [reply]
Right, `org.graalvm.nativeimage` is also part of the `nativeimage.jar`

**Comment by chumer** [comment]
graal-sdk as of 23.1 is deprecated. We still deploy a module with transitive requires to maven for compatibility. We no longer deploy it as part of any graalvm jdk. Graalvm jdks only contain the org.graalvm.nativeimage module (and it's dependencies). I assume that we will stop deploying graal-sdk.jar in 24.0 or a later release if there is no demand to keep it.

All details are described as usual in the changelog:
https://github.com/oracle/graal/blob/master/sdk/CHANGELOG.md#version-2310

I cannot comment in more detail on com.oracle.svm.core.annotate. The native-image team should comment. For me it's a package with a bad name that is part of the native-image module mostly for compatibility. My guess would be that it gets phased out over time with a replacement that better fits the other packages of the module.

**Reply by jerboaa** [reply]
>All details are described as usual in the changelog:
>https://github.com/oracle/graal/blob/master/sdk/CHANGELOG.md#version-2310

This suggests the API jars (Maven modules) with GraalVM for JDK 21 (`23.1`) will be the following:

```
org.graalvm.sdk:nativeimage
org.graalvm.sdk:word
org.graalvm.sdk:collections
```

and (whoever needs it):

```
org.graalvm.polyglot:polyglot
```

I take it those `org.graalvm.sdk` modules above will remain stable in terms of classes they contain (API surface) and compatible JDK version a la https://github.com/oracle/graal/issues/6831? Could somebody please confirm/deny?

**Reply by chumer** [reply, marked as answer]
> I take it those org.graalvm.sdk modules above will remain stable in terms of classes they contain (API surface) and compatible JDK version a la https://github.com/oracle/graal/issues/6831? Could somebody please confirm/deny?

Yes. We do not do breaking changes in these packages (without long-term deprecation). Technically this change is a compatible change, as the org.graalvm.sdk still provides the same packages classes it did before, unless you did break the encapsulation of an individual package with --add-exports then this will be a breaking change as you need to update the module-name of the package.

**Reply by jerboaa** [reply]
Thanks.

**Comment by christianwimmer** [comment]
> I cannot comment in more detail on com.oracle.svm.core.annotate. The native-image team should comment. 

Nothing is going to change regarding the com.oracle.svm.core.annotate package. It will remain in the same module/jar where the rest of the native-image API is.

> com.oracle.svm.core.annotate (which AFAIK is public API)

No, it is not API as the JavaDoc clearly states: https://github.com/oracle/graal/blob/master/sdk/src/com.oracle.svm.core.annotate/src/com/oracle/svm/core/annotate/package-info.java#L47
But for pragmatic reasons, we are also not going to change or remove it.

**Reply by jerboaa** [reply]
OK, noted. Thanks!
